File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.java
import java.util.*;
import java.lang.*;
import java.io.*;

class Solution {
    HashMap<String, String> hm = new HashMap<String, String>();
    ArrayList<String[]> arr = new ArrayList<>();
    
    public String[] solution(String[] record) {
        String[] answer = {};
        ArrayList<String> res = new ArrayList<String>();
        
        for(int i = 0; i < record.length; i++){
            StringTokenizer st = new StringTokenizer(record[i]);
            ArrayList<String> tmp = new ArrayList<String>();
            while(st.hasMoreTokens()){
                tmp.add(st.nextToken());
            }
            
            String ac = tmp.get(0);
            String id = tmp.get(1);
            String name;
            if(!(tmp.get(0).equals("Leave"))){
                name = tmp.get(2);
                hm.put(id, name);
            }
            
            if(!(tmp.get(0).equals("Change"))){
                arr.add(new String[]{ac, id});
            }
        }
        
        for(int i = 0; i < arr.size(); i++){
            String tmp = "";
            
            tmp += hm.get(arr.get(i)[1]) + "님이 ";
            
            if(arr.get(i)[0].equals("Enter")){
                tmp += "들어왔습니다.";
            }
            else
                tmp += "나갔습니다.";
            
            res.add(tmp);
        }
        return res.toArray(new String[0]);
    }
}

File: 프로그래머스/2/42888. 오픈채팅방/오픈채팅방.cpp
#include <string>
#include <vector>
#include <map>
#include <sstream>
#include <iostream>

using namespace std;

typedef struct _data{
    string ac;
    string id;
}DATA;

vector<string> solution(vector<string> record) {
    vector<string> answer;
    map<string, string> m;
    vector<DATA> r;
    for(int i = 0; i < record.size(); i++){
        string tmp = record[i];
        
        stringstream ss(tmp);
        string word;
        vector<string> words;
        
        while(ss >> word){
            words.push_back(word);    
        }
        
        string ac, id, name;
        
        ac = words[0];
        id = words[1];
        
        if(ac != "Leave"){
            name = words[2];
        }
        
        
        if(ac != "Change")
            r.push_back({ac, id});
        
        if(ac == "Enter" || ac == "Change")
            m[id] = name;
    }
    
    for(int i = 0; i < r.size(); i++){
        string tmp = "";
        tmp += m[r[i].id];
        tmp += "님이 ";
        if(r[i].ac == "Enter"){
            tmp += "들어왔습니다.";
        }
        else{
            tmp += "나갔습니다.";
        }
        answer.push_back(tmp);
    }
    
    return answer;
}